File: app/src/main/java/com/obsidian/plugins/task_notifier/os/FileObserverService.java
package com.obsidian.plugins.task_notifier.os;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.FileObserver;
import android.os.IBinder;

import androidx.documentfile.provider.DocumentFile;

import com.obsidian.plugins.task_notifier.core.ObsidianTaskReminderCore;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class FileObserverService extends Service {
    private FileObserver mFileObserver;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Context context = this;
        System.out.println("service called");

        if ((intent.hasExtra(ServiceManager.FILE_PATH_INTENT_KEY))) // we store the path of directory inside the intent that starts the service
            System.out.println("service called inside");
        String path = intent.getStringExtra(ServiceManager.FILE_PATH_INTENT_KEY);
        mFileObserver = new FileObserver(path) {
            @Override
            public void onEvent(int event, String path) {
                if (path == null) return;
                if (!path.endsWith(".md")) return;
                if (event != MODIFY) return;
                String changeContent = path;

                Uri myUri = Uri.parse("content://com.android.externalstorage.documents/tree/primary%3Atest%2Ftest");
                try {
                    DocumentFile document = DocumentFile.fromTreeUri(context, myUri);
                    for (DocumentFile file : document.listFiles()) {
                        if (!file.isFile()) continue;
                        if (!file.getName().contains(path)) continue;
                        InputStream fis = context.getContentResolver().openInputStream(file.getUri());
                        InputStreamReader isr = new InputStreamReader(fis);
                        BufferedReader bufferedReader = new BufferedReader(isr);
                        StringBuilder sb = new StringBuilder();
                        String line;
                        while ((line = bufferedReader.readLine()) != null) {
                            sb.append(line);
                        }
                        fis.close();
                        changeContent = sb.toString();
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
                ObsidianTaskReminderCore.onFileChanged(context, path);

            }
        };
        mFileObserver.startWatching(); // The FileObserver starts watching
        return Service.START_NOT_STICKY;
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public IBinder onBind(Intent intent) {
        //TODO for communication return IBinder implementation
        return null;
    }
}

File: app/src/main/java/com/obsidian/plugins/task_notifier/os/FIleObserverService.kt
package com.obsidian.plugins.task_notifier.os

import android.app.Service
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.FileObserver
import android.os.IBinder
import com.obsidian.plugins.task_notifier.core.ObsidianTaskReminderCore
import com.obsidian.plugins.task_notifier.core.OnFileChangedResult
import com.obsidian.plugins.task_notifier.utils.Constants
import com.obsidian.plugins.task_notifier.utils.FileUtils
import com.obsidian.plugins.task_notifier.utils.Logger
import com.obsidian.plugins.task_notifier.utils.ScopeEnum
import java.io.File

class FileObserverService : Service() {
  private var mFileObserver: FileObserver? = null
  private var filePath: String? = null
  private var fileUri: Uri? = null

  override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
    Logger.info("FileObserverService.onStartCommand")
    val context: Context = this
    val uriString = intent.getStringExtra(ServiceManager.FILE_PATH_INTENT_KEY)
    if (uriString == null) {
      Logger.info("FileObserverService missing path from FILE_PATH_INTENT_KEY intent key")
      return STOP_FOREGROUND_REMOVE
    }

    Logger.info("FileObserverService started for uri ${uriString}")
    val uri = Uri.parse(uriString)
    fileUri = uri
    mFileObserver = getFileObserver(context, uri)
    (mFileObserver as FileObserver).startWatching() // The FileObserver starts watching

    val changeContent = FileUtils().tryReadFileContent(context, fileUri!!)

    ObsidianTaskReminderCore.onFileChanged(
      context,
      fileUri!!,
      changeContent
    )
    val notification = NotificationManager.notify(
      context,
      Constants.APPLICATION_NAME,
      "👀 Watching: $filePath",
      hashCode(),
      ScopeEnum.APPLICATION
    )
    startForeground(1, notification)
    return START_STICKY
  }

  override fun onBind(p0: Intent?): IBinder? {
    Logger.info("FileObserverService.onBind")
    TODO("Not yet implemented")
  }

  override fun onCreate() {
    super.onCreate()
    Logger.info("FileObserverService.onCreate")
  }

  override fun onDestroy() {
    super.onDestroy()
    Logger.info("FileObserverService.onDestroy")
  }

  private fun getFileObserver(context: Context, uri: Uri): FileObserver {
    val realPath = FileUtils().getPath(context, uri)
    Logger.info("FileObserverService created for file: $realPath ")
    filePath = realPath
    return object : FileObserver(File(realPath), FileObserver.CLOSE_WRITE) {
      override fun onEvent(event: Int, path: String?) {
        Logger.info("FileObserverService.onEvent $event path: $realPath")

        var changeContent = FileUtils().tryReadFileContent(context, fileUri!!)

        val fileChangeACKResult = ObsidianTaskReminderCore.onFileChanged(
          context,
          fileUri!!,
          changeContent
        )
        if (fileChangeACKResult === OnFileChangedResult.STOP_LISTENING) {
          Logger.info("FileObserverService stopping listening!")
          mFileObserver!!.stopWatching()
        }
      }
    }
  }
}
